homeassistant/components/opentherm_gw/__init__.py

```diff
@@ -75,13 +135,111 @@
     await gateway.connect(hass.loop, device_path)
     _LOGGER.debug("Connected to OpenTherm Gateway at %s", device_path)
 
+    async def cleanup(event):
+        """Reset overrides on the gateway."""
+        await gateway.set_control_setpoint(0)
+        await gateway.set_max_relative_mod('-')
+    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, cleanup)
+
     async def handle_report(status):
         """Handle reports from the OpenTherm Gateway."""
         _LOGGER.debug("Received report: %s", status)
+        hass.data[DATA_OPENTHERM_GW][DATA_LATEST_STATUS] = status
         async_dispatcher_send(hass, SIGNAL_OPENTHERM_GW_UPDATE, status)
     gateway.subscribe(handle_report)
 
 
+async def register_services(hass, gateway):
+    """Register services for the component."""
+    gw_vars = hass.data[DATA_OPENTHERM_GW][DATA_GW_VARS]
+
+    async def reset_gateway(call):
+        """Reset the OpenTherm Gateway."""
+        mode_rst = gw_vars.OTGW_MODE_RESET
+        status = await gateway.set_mode(mode_rst)
+        hass.data[DATA_OPENTHERM_GW][DATA_LATEST_STATUS] = status
+        async_dispatcher_send(hass, SIGNAL_OPENTHERM_GW_UPDATE, status)
+    hass.services.async_register(DOMAIN, SERVICE_RESET_GATEWAY, reset_gateway)
+
+    async def set_control_setpoint(call):
+        """Set the control setpoint on the OpenTherm Gateway."""
+        gw_var = gw_vars.DATA_CONTROL_SETPOINT
+        value = await gateway.set_control_setpoint(call.data[ATTR_TEMPERATURE])
+        status = hass.data[DATA_OPENTHERM_GW][DATA_LATEST_STATUS]
+        status.update({gw_var: value})
+        async_dispatcher_send(hass, SIGNAL_OPENTHERM_GW_UPDATE, status)
+    hass.services.async_register(DOMAIN, SERVICE_SET_CONTROL_SETPOINT,
+                                 set_control_setpoint,
+                                 SERVICE_SET_CONTROL_SETPOINT_SCHEMA)
+
+    async def set_device_clock(call):
+        """Set the clock on the OpenTherm Gateway."""
+        attr_date = call.data[ATTR_DATE]
+        attr_time = call.data[ATTR_TIME]
+        hass.async_add_job(gateway.set_clock, datetime.combine(attr_date,
```

Comment:
Like the other commands/services it's a coroutine that waits for a response from the gateway, but we don't need the return value here. That's why i awaited the others and added this one as an async job. Should I use another function instead of async_add_job?

Reply:
Assuming there's no reason no to wait for the result I mean.

Reply:
There is indeed no reason not to wait for the result as set_device_clock is already running in the event loop and it's the last statement in this function. I will change it to an await statement to avoid the overhead of async_add_job.